Error writing maps with exactly 64 entries

It took me a while to figure out this particular problem. Even after #10 I still have trouble writing a small subset of our data, and I was suprised to narrow it down to dictionaries with exactly 64 elements. Here's a minimal example:

```
julia> Avro.write([Dict("$i" => i for i in 1:64)])
ERROR: AssertionError: nb == pos - startpos
Stacktrace:
 [1] writevalue(B::Avro.Binary, A::Avro.ArrayType, x::Vector{Dict{String, Int64}}, buf::Vector{UInt8}, pos::Int64, len::Int64, opts::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Avro ~/.julia/dev/Avro/src/types/arrays.jl:44
 [2] write(obj::Vector{Dict{String, Int64}}; schema::Avro.ArrayType, jsonencoding::Bool, kw::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Avro ~/.julia/dev/Avro/src/types/binary.jl:26
 [3] write(obj::Vector{Dict{String, Int64}})
   @ Avro ~/.julia/dev/Avro/src/types/binary.jl:22
 [4] top-level scope
   @ REPL[51]:1
```

It works fine with 63 or 65 entries. Like #9, the number of bytes written is not match `nbytes`. The value of 64 is suggestive but I don't see an obvious problem with that:

```
julia> Avro.nbytes(64)
2

julia> Avro.write(64)
2-element Vector{UInt8}:
 0x80
 0x01

julia> Avro.read([0x80, 0x01], Int)
64
```

Searching all lengths exhaustively from `0:2^15` I find the same issue at 8192. This is another point where the integer adds a byte, `[0x80, 0x80, 0x01]`. I get the same error at 2^20, so there is a pattern here. Does this suggest anything to you @quinnj ?

--- a/src/types/arrays.jl
+++ b/src/types/arrays.jl
@@ -49,7 +49,7 @@ end
 
 function nbytes(A::ArrayType, x, self=false)
     nb = isempty(x) ? 0 : sum(y -> nbytes(A.items, y), x)
-    return self ? nb : nbytes(length(x)) + nbytes(nb) + nb + 1
+    return self ? nb : nbytes(-length(x)) + nbytes(nb) + nb + 1
 end
 
 function readvalue(B::Binary, AT::ArrayType, ::Type{A}, buf, pos, buflen, opts) where {A <: AbstractVector{T}} where {T}

--- a/src/types/maps.jl
+++ b/src/types/maps.jl
@@ -44,7 +44,7 @@ function nbytes(MT::MapType, x, self=false)
         nb += nbytes(MT.values, v)
         next = iterate(pairs, state)
     end
-    return nb == 0 ? 1 : self ? nb : nbytes(length(x)) + nbytes(nb) + nb + 1
+    return nb == 0 ? 1 : self ? nb : nbytes(-length(x)) + nbytes(nb) + nb + 1
 end
 
 function readvalue(B::Binary, MT::MapType, ::Type{A}, buf, pos, buflen, opts) where {A <: AbstractDict{K, V}} where {K, V}

